### PR TITLE
Enhancement/issues#9 add angular karma setup

### LINKED ISSUE #9: Add Angular2 basic setup to starter template.
Add Angular2 library.
Add basic test
Make sure styles work
Modify webpack configs accordingly.

### PR DESCRIPTION
**Pull request by cxoltero**
resolves #9 , #3 , #6 
Started a clean branch since too much cleanup needed in other pr.
See comments below.

### CODE REVIEW

File: webpack.config.common.js [hunk 1]
```diff
@@ -76,8 +81,13 @@ module.exports = {
...
-      filename: 'styles.css',
+      filename: '[name].css',
       allChunks: true
-    })
+    }),
+
+    new webpack.ContextReplacementPlugin(
+      /angular(\\|\/)core(\\|\/)@angular/,
+      path.resolve(__dirname, 'src')
+    )
```
Comment (cxoltero) [inline webpack.config.common.js]: Issue:  console warning
Documented on: [#11580](https://github.com/angular/angular/issues/11580)
Log: /node_modules/@angular/core/@angular/core.es5.js Critical dependency: the request of a dependency is an expression
Fix: add the following to typescript plugins
```
new webpack.ContextReplacementPlugin(
 /angular(\\|\/)core(\\|\/)@angular/,
 path.resolve(__dirname, 'src')
)
```

File: src/index.ts [hunk 2]
```diff
@@ -1,8 +1,20 @@
...
+import { NgModule, enableProdMode } from '@angular/core';
+import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
+import { BrowserModule } from '@angular/platform-browser';
 
-    app.innerHTML = '<h1>' + title + '</h1>';
+import { BootstrapComponent } from './components/bootstrap/bootstrap.component';
+
+if (process.env.ENV === 'production') {
+  enableProdMode();
 }
```
Comment (cxoltero) [inline src/index.ts]: enableProdMode -  turns off assertions and other checks within the framework.

File: webpack.config.common.js [hunk 3]
```diff
@@ -43,17 +49,16 @@ module.exports = {
...
-        exclude: path.join(__dirname, './index.html')
+        loader: 'html-loader'
       }, {
         test: /\.(s)css/,
-        use: ExtractTextPlugin.extract({
-          fallback: 'style-loader',
-          use: 'css-loader!sass-loader'
-        })
+        loader: 'raw-loader!sass-loader'
       }, {
```
Comment (cxoltero) [inline webpack.config.common.js]: When working with node-sass i randomly ran into the following error:
Log: Error: ENOENT: no such file or directory, scandir
Fix: delete node_modules directory, then run yarn install
Refference: [#1579](https://github.com/sass/node-sass/issues/1579)

File: karma.config.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+webpackConfig.plugins[2] = new CommonsChunkPlugin({
+  name: 'tests/common.js',
+  minChunks: function ({ resource }) {
+    return resource &&
+      resource.indexOf('node_modules') >= 0 &&
+      resource.match(/\.js$/);
+  }
+});
+
```
Comment (cxoltero) [inline karma.config.js]: Issue: karma-webpack does not exclude vendor libs from test bundles when using commonChunkPlugin
Possible solutions: 
1.- Create separate webpack-config-test.js without CommonsChunkPlugin
	sample : https://angular.io/guide/webpack#test-configuration
2.- Create workaround shim as shown in: [karma-webpack#issue22](https://github.com/webpack-contrib/karma-webpack/issues/22#issuecomment-313827844), this workaround imports all necessary libraries via a common.js file inside of tests directory. then we fetch those files inside our karma config.

File: webpack.config.common.js [hunk 5]
```diff
@@ -67,7 +72,7 @@ module.exports = {
     }),
 
     new webpack.optimize.CommonsChunkPlugin({
-      name: 'common'
+      name: ['index', 'vendor', 'polyfills']
     }),
```
Comment (cxoltero) [inline webpack.config.common.js]: angular docs recommend splitting libraries into vendor chunkfile.

Comment (thescientist13) [inline webpack.config.common.js]: can you please provide a link and your observations on doing it this way vs the example we committed to in [react-webpack-stack](https://github.com/kenzanlabs/react-webpack-stack/issues/23)

My understanding of the current approach was that this
```
new webpack.optimize.CommonsChunkPlugin({
      name: 'common'
    }),
```

Would take any "common" code observed by webpack as it built the AST and put them into a new file, in this case called common.  Are you saying this does the same but in a different way?   Is it a better way to introduce this deviation, and thus the recommendation would be to backport it to react-webpack-stack?

cc: @thisgeek  /  @paul-barry-kenzan

File: karma.config.js (hunk 4 above)
Comment (thescientist13) [inline karma.config.js]: please include a link to that github issue.  

and also, just confirming that [this](https://github.com/thegreenhouseio/angular-webapp-seed/blob/master/karma.conf.js#L7) doesn't work?  
```
// TODO issues with karma and CommonChunksPlugin
// https://github.com/webpack/karma-webpack/issues/24
webpackConfig.plugins[0] = function() {};
```

File: src/vendor.ts [hunk 6]
```diff
@@ -1 +1,15 @@
...
+import '@angular/common';
+import '@angular/http';
+import '@angular/router';
+
+// RxJS
+import 'rxjs';
+
+// Other vendors for example jQuery, Lodash or Bootstrap
+// You can import js, ts, css, sass, ...
+
```
Comment (thescientist13) [inline src/vendor.ts]: we can probably remove the `console` here

File: tsconfig.json [hunk 7]
```diff
@@ -1,18 +1,32 @@
...
+    "noImplicitAny": true,
     "outDir": "./build/",
+    "removeComments": false,
     "sourceMap": true,
-    "noImplicitAny": true,
-    "module": "commonjs",
-    "target": "es2015",
-    "allowJs": true,
-    "allowSyntheticDefaultImports": true
+    "suppressImplicitAnyIndexErrors": true,
```
Comment (thescientist13) [inline tsconfig.json]: I still would like to get an acknowledgement of what we're adding here that we didn't need to add to [react-webpack-stack](https://github.com/kenzanlabs/react-webpack-stack/blob/master/tsconfig.json) for the article documentation

File: webpack.config.common.js [hunk 8]
```diff
@@ -8,14 +8,15 @@ const webpack = require('webpack');
...
     vendor: './vendor.ts'
   },
 
   output: {
-    path: path.join(__dirname, './build'),
-    filename: '[name].bundle.js',
-    sourceMapFilename: '[name].map',
-    chunkFilename: '[id].[name]chunk.js'
+    path: path.join(__dirname, 'build'),
+    filename: '[name].[hash].bundle.js',
```
Comment (thescientist13) [inline webpack.config.common.js]: why not `[chunkhash]` like in [react-webpack-stack](https://github.com/kenzanlabs/react-webpack-stack/blob/master/webpack.config.common.js#L13)?

File: webpack.config.common.js (hunk 5 above)
Comment (thescientist13) [inline webpack.config.common.js]: After reverting back to what was there before, not only do we know have our "common" chunk, but our overall bundle sizes went down a bit too.
screen shot 2017-08-08 at 9 48 42 am

File: src/vendor.ts [hunk 9]
```diff
@@ -1 +1,15 @@
+// Angular
+import '@angular/platform-browser';
+import '@angular/platform-browser-dynamic';
+import '@angular/core';
+import '@angular/common';
+import '@angular/http';
+import '@angular/router';
+
+// RxJS
+import 'rxjs';
```
Comment (thescientist13) [inline src/vendor.ts]: importing all of *rxjs* is a bad practice, since the library is quite large and probably accounts for most of the 2.8MB in _vendor.ts_ .  it is recommended to only import the operators needed for your application
https://stackoverflow.com/questions/42376972/best-way-to-import-observable-from-rxjs

File: webpack.config.common.js (hunk 8 above)
Comment (cxoltero) [inline webpack.config.common.js]: We should not use [chunkhash] or [hash] for development. This will cause many other issues, like a memory leak, because the dev server does not know when to clean up the old files.
See [webpack-dev-server#377](https://github.com/webpack/webpack-dev-server/issues/377). Creating ticket/issue for [react-webpack-stack](https://github.com/kenzanlabs/react-webpack-stack/blob/master/webpack.config.common.js#L13)

**Comment by thescientist13** [comment]
@cxoltero 
Please review these changes [1](https://github.com/kenzanlabs/angular-webpack-stack/pull/12/commits/ad7460becd88eb0f833e305e6fea96789a3f2b68) and [2](https://github.com/kenzanlabs/angular-webpack-stack/pull/12/commits/ecf1caf29d1b20a18dd4ed90017873b182caa4e5)

Summary of Changes
1.  was able to clean up Karma configuration override of CommonsChunkPlugin (like [here](https://github.com/thegreenhouseio/angular-webapp-seed/blob/master/karma.conf.js#L7))
1.  One less file to track of by keeping all file includes in _karma.config.js_ (see [here](https://github.com/thegreenhouseio/angular-webapp-seed/blob/master/karma.conf.js#L17)
1.  Aligned `output` configuration with react-webpack-stack
1.  Moved **HotModuleReplacementPlugin** to develop config
1.  Removed `es6-shim` as it appeared there was overlap between it and **core-js**
1.  Removed a lot of configuration from _tsconfig.json_ to keep it much more aligned with react-webpack-stack

Article Takeaways (for me)
1.  Document **ContextReplacementPlugin** (added to _webpack.config.common.js_)
1.  Document need for `emitDecoratorMetadata` and `experimentalDecorators` which are specific to Angular's usage of Decorators (and was also pretty obvious from the console errors)
1.  Aligned usage of `ExtractTextPlugin` to better match react-webpack-stack, leaving a default implementation in _webpack.config.common.js_
1.  I made a ticket to track making sure HMR works - #14
1.  I made a ticket to track only importing what's needed from **rxjs** (not priority) - #13

In summary, I found that there was a lot of configuration, like in _tsconfig.json_ we didn't need.  My hope is that in particular when it comes to configuration (for this or any other project) that you only add what you need when you need it.  Start small and work your way up, enabling things as they make errors go away.

Let me know your thoughts, but otherwise I think this is good to go.

**Comment by cxoltero** [comment]
Yes, also a lot of that overhead configuration and bugs came from putting hot reload in the webpack.config file instead of the webpack.dev.
